apps/prairielearn/src/lib/file-store.ts

```diff
@@ -120,79 +131,82 @@ export async function deleteFile(file_id, authn_user_id) {
 /**
  * Option of returning a stream instead of a file
  *
- * @param {number | string} file_id - The file to get.
- * @return {Promise<(import('stream'))>} - Requested file stream.
+ * @param file_id - The file to get.
+ * @return Requested file stream.
  */
-export async function getStream(file_id) {
+export async function getStream(file_id: number | string): Promise<Stream> {
   debug(`getStream(): file_id=${file_id}`);
   const file = await getFile(file_id, 'stream');
   return file.contents;
 }
 
+export async function getFile(
+  file_id: number | string,
+  data_type: 'stream',
+): Promise<{
+  contents: Stream;
+  file: File;
+}>;
+
+export async function getFile(
+  file_id: number | string,
+  data_type?: 'buffer',
+): Promise<{
+  contents: Buffer;
+  file: File;
+}>;
 /**
  * Get a file from the file store.
  *
- * @param {number | string} file_id - The file to get.
- * @return {Promise<object>} An object with a buffer (of the file contents) and a file object.
+ * @param file_id - The file to get.
+ * @return An object with a buffer (of the file contents) and a file object.
  */
-export async function getFile(file_id, data_type = 'buffer') {
-  debug(`get(): file_id=${file_id}`);
-  const params = { file_id };
-  const result = await sqldb.queryZeroOrOneRowAsync(sql.select_file, params);
-  debug('get(): got row from DB');
-
-  let buffer, readStream;
-
-  if (result.rows.length < 1) {
+export async function getFile(
```

Comment:
I was really hoping we could type this function such that it'd actually be safe (that is, if the caller specifies `'buffer'`, the type system would guarantee that a Buffer is returned, same for streams) but that doesn't seem possible: I deliberately broke the function by changing the buffer branch to return a stream and the compiler didn't say anything. I'm wondering if we should just use two functions: `getFileBuffer` and `getFileStream`? Same for S3 maybe? FWIW, there aren't any callers outside of this file using the stream variant (those use `getStream`, which is a) named poorly and b) just a wrapper around this function).

I won't block on this.

Reply:
I was not able to observe this.

![CleanShot 2024-12-23 at 13 22 27@2x](https://github.com/user-attachments/assets/ee8f1916-3212-4fd8-9153-7e25aba0d4a8)
![CleanShot 2024-12-23 at 13 22 43@2x](https://github.com/user-attachments/assets/8601ec49-c04e-4695-8c29-5107481f5bcf)

Reply:
You're testing something different from what I described above. Yes, the types on `getFile` work correctly for callers, but they don't actually help us enforce that `getFile` itself is implemented correctly. To be more specific about what I tried above: in the _implementation_ of `getFile`, I changed a `data_type === 'buffer'` branch to return a stream and the compiler didn't complain.